### DISCUSSION TITLE
Possible to use multiple message files per one locale?

### DISCUSSION THREAD

**Discussion by Doesntmeananything** [amannn/next-intl#357] [Q&A] [answered]
Looking at the [Messages](https://next-intl-docs.vercel.app/docs/usage/messages) section of the docs, it's not quite clear if it's possible to use multiple message files per one locale.

What I'm wondering is if it's possible to do something like this:
```
── messages
    └── en
        ├── login.json
        └── dashboard.json
    └── fr
        ├── login.json
        └── dashboard.json
```

Would this require additional middleware customization? And I'm not sure how that would work with TypeScript and passing namespaces to `useTranslations()`, e.g. `useTranslations('dashboard')`.

Thanks!

**Comment by amannn** [comment, marked as answer]
I'm currently working on a [major docs update](https://github.com/amannn/next-intl/discussions/351) and have added a section that should hopefully answer this: [Messages configuration](https://next-intl-docs.vercel.app/docs/usage/configuration#messages) ("How can I split my messages into multiple files?").

Feel free to have a look! If you have any feedback, please let me know in https://github.com/amannn/next-intl/discussions/351

**Reply by Doesntmeananything** [reply]
That looks like exactly what I wanted to know, thank you very much!

Appreciate all the hard work you're putting in.

**Reply by amannn** [reply]
On a related note, there's now a VSCode integration for `next-intl` which helps to streamline the management of translations in large files: [VSCode integration docs](https://next-intl-docs.vercel.app/docs/workflows/vscode-integration).

**Reply by lhguerra** [reply]
I saw the linked docs but still don't have a clue on how to split the locale into several files. What am I missing?

**Reply by IngEoGeoBuk** [reply]
Hi. There!
I also tried to split my messages into multiple files, but it doesn't work, 
recurring an error 'Error: No messages found. Have you configured them correctly?' when use useMessages() into layout.tsx

My code: https://github.com/IngEoGeoBuk/nextjs14-next-intl-tutorial/tree/split-multiple-files
(plz check the branch name 'split-multiple-files')
<img width="1876" alt="screenshot" src="https://github.com/amannn/next-intl/assets/59249899/4e28c3b9-5fdc-4f7e-8419-f6c4d27409df">

I modified only i18n.ts and json messages files. 
middleware.ts and layout.tsx, next.config.mjs are same as https://next-intl-docs.vercel.app/docs/getting-started/app-router

How to solve it? Thank you.

**Reply by IngEoGeoBuk** [reply]
Hi. There! I found the answer.

change i18n.ts like this

```
import { notFound } from "next/navigation";
import { getRequestConfig } from "next-intl/server";

// Can be imported from a shared config
const locales = ["en", "ko"];

export default getRequestConfig(async ({ locale }) => {
  if (!locales.includes(locale as any)) notFound();

  const about = await import(`../messages/${locale}/about.json`);
  const home = await import(`../messages/${locale}/home.json`);
  const room = await import(`../messages/${locale}/room.json`);
  return {
    messages: {
      ...about,
      ...home,
      ...room,
    },
  };
});
```

**Reply by SalahAdDin** [reply]
`locales as any`?

**Reply by quyphan97** [reply]
@IngEoGeoBuk Hi! can you give me an example of multiple file message + localePrefix: "as-needed"? I'm having trouble getting translate to work using "use client" in components. :(

**Comment by viclafouch** [comment]
On my side, I use an another strategie, and it's working well, but instead of "json" file, it's `ts` extension.

Here my folder structure: 

<img width="384" alt="image" src="https://github.com/amannn/next-intl/assets/23353836/e42b1077-d615-4593-8b6d-239833f0a4bf">

```ts
// fr/common.fr
export default {
  hello: 'Bonjour'
} as const
```

```ts
// fr/index.ts
import common from './common'

// "common" will be used as a namespace
export default { common }
```


```ts
// global.d.ts
// Use type safe message keys with `next-intl`

type Messages = (typeof import('./locales/en'))['default']

declare interface IntlMessages extends Messages {}
```

```ts
// index.ts (no need changes when you add a namespace file)
import { notFound } from 'next/navigation'
import { getRequestConfig } from 'next-intl/server'
import { locales } from '@i18n/config'

export default getRequestConfig(async ({ locale }: { locale: string }) => {
  if (!locales.includes(locale)) {
    notFound()
  }

  const messages = await import(`./locales/${locale}/index.ts`)

  return {
    messages: messages.default,
    timeZone: 'Europe/Vienna',
    now: new Date()
  }
})

```

Hope it helps !

**Reply by bracebrace008** [reply]
json is better, can be intergrated to another none js project

**Reply by SalahAdDin** [reply]
@bracebrace008 @Doesntmeananything  It does not work right with types.

**Comment by dogukankaynar** [comment]
Below I explain the way I found in the json structure in a Type safe way
Our folder structure:    
```javascript 
messages
├── login
│   ├── en.json
│   └── tr.json
└── profile
    ├── en.json
    └── tr.json
```

 
 
```javascript
// i18n.ts

import { getRequestConfig } from 'next-intl/server'


export default getRequestConfig(async ({ locale }) => {
  const loginMessageImport = await import(`./messages/login/${locale}.json`);
  const loginMessages = loginMessageImport.default;
  
  const profileMessageImport = await import(`./messages/profile/${locale}.json`);
  const profileMessages = profileMessageImport.default;
  //Importing other language files ..


// Combine all messages into a single object

  const messages = {
    ...loginMessages,
    ....profileMessages,
    // Add messages from other language files here
  };

  // Return merged messages
  return messages;

});
```

```javascript
// global.d.ts

type LoginMessages = typeof import("./messages/login/en.json");
type ProfileMessages = typeof import("./messages/profile/en.json");
// Importing other language files ..

// Create a new type by combining all message types
type Messages = LoginMessages & ProfileMessages;

declare interface IntlMessages extends Messages {}

```

**Reply by SalahAdDin** [reply]
So, for global type, it is only enough to import files from one language to get the types correctly, right?

**Reply by dogukankaynar** [reply]
yes if key and value are equal in two json files

**Reply by SalahAdDin** [reply]
It is failing for me:
![image](https://github.com/amannn/next-intl/assets/5034215/8d92ee11-b021-419a-84f3-808f06d32014)
But if we directly define the locale, it works:
![image](https://github.com/amannn/next-intl/assets/5034215/4f16ad1e-9ea7-4f2a-b8c2-ef60454e83dc)

The same happens when using my old configuration: 
![image](https://github.com/amannn/next-intl/assets/5034215/250468e6-b56c-4d2f-ba86-efe855434a48)

**Reply by SalahAdDin** [reply]
Using `as` works:
![image](https://github.com/amannn/next-intl/assets/5034215/374fdac9-55c0-42b7-a09e-ce7f26871c3a)

**Reply by SalahAdDin** [reply]
@amannn what do you think about this?

**Reply by amannn** [reply]
Sure, looks ok!

**Reply by SalahAdDin** [reply]
I mean, it does not solve the typing issue.

**Comment by victorvianaom** [comment]
Hi @amannn , thanks for the great work.
For the app router, given a locale, every page will receive the same messages?
Is it possible to send diferent namespaces per route like we do for the pages router?

**Reply by amannn** [reply]
Yep, you can, but unfortunately this is arguably a bit more involved with the App Router: https://next-intl-docs.vercel.app/docs/environments/server-client-components#option-3-providing-individual-messages.

There's work happening on an automated way to do this in the future: https://github.com/amannn/next-intl/issues/1 🤞

**Reply by victorvianaom** [reply]
Thanks for the response @amannn ! I understood the intricacies of it.

**Comment by Christian-Hoeller** [comment]
I have created this config in the pages router for a more dynamic approach


Function that loads the config
```typescript
// i18n.ts

const translationsConfig = [
  {
    page: "index",
    files: ["index", "pricing"],
  },
];

export const loadLangFiles = async ({ page = "index", locale = "en" }) => {
  const translationFilesNames = translationsConfig.find(
    (item) => item.page === "index",
  )?.files;

  if (!translationFilesNames)
    throw new Error(`Translation configuration for ${page} does not exist!`);

  const translationObjects = await Promise.all(
    translationFilesNames.map((file) =>
      loadLangFile({
        file,
        locale,
      }),
    ),
  );
  return Object.assign({}, ...translationObjects) as IntlMessages;
};

export const loadLangFile = async ({ file = "index", locale = "en" }) => {
  return await import(`../messages/${locale}/${file}.json`);
};

```

Use in your page like this:
```typescript
// index.ts

export async function getStaticProps(context: GetStaticPropsContext) {
  return {
    props: {
      messages: await loadLangFiles({}),
    },
  };
}
```

TypeSafety:
```typescript
// global.d.ts

type IndexMessages = typeof import("../messages/en/index.json");
type PricingMessages = typeof import("../messages/en/pricing.json");

type Messages = IndexMessages & PricingMessages;

declare interface IntlMessages extends Messages {}

**Comment by HPouyanmehr** [comment]
You can also use [`next-intl-split`](https://github.com/HPouyanmehr/next-intl-split) if you like to separate everything properly, at the end of the day it'll merge all translations into a single messages json and you can use the namespace as well.

**Reply by SalahAdDin** [reply]
> You can also use [`next-intl-split`](https://github.com/HPouyanmehr/next-intl-split) if you like to separate everything properly, at the end of the day it'll merge all translations into a single message `json` and you can use the namespace as well.

An extra package? I should be a native feature I think.

**Reply by HPouyanmehr** [reply]
Hey @SalahAdDin, thanks for your feedback.
Generally, I prefer to have this ability in the current package same as you 🤝. 
The current built-in solution to separate the translation files can lead to name conflicts (1),  also you need to add an import statement for every new file (2), and you may lose the benefits of `namespace` (3).

`next-inlt-split` is a tiny package, with a cost of 2Kb, that can save you some time, I disagree with calling that a "negative feature", I see that as an extra solution for this topic. In my book, it depends on the project scope and the developer's available time to see if adding a new package is worth it. 

![next-intl-split-import-cost](https://github.com/user-attachments/assets/76e378c1-8141-471e-b0c0-3c283c241e99)

**Reply by SalahAdDin** [reply]
> `next-inlt-split` is a tiny package, with a cost of 2Kb, that can save you some time, I disagree with calling that a "negative feature", I see that as an extra solution for this topic. In my book, it depends on the project scope and the developer's available time to see if adding a new package is worth it.
> 

It is a good feature, it should be included in the package by default!

**Reply by aramix** [reply]
@HPouyanmehr  How can it be used with `next-config.mjs`

**Reply by brandanking-decently** [reply]
@amannn what is the likely hood of something like next-intl-split being baked into the package. Would be great to have something like this directly that can also generate the correct types for the useTranslation hook

**Reply by SalahAdDin** [reply]
> @amannn what is the likely hood of something like next-intl-split being baked into the package. Would be great to have something like this directly that can also generate the correct types for the useTranslation hook

I agree with you; @amannn  what do you think?

**Reply by amannn** [reply]
`next-intl` is currently agnostic to how you load your messages: It can be a single file, you can fetch from an endpoint, use multiple files etc. I think this approach gives a lot of flexibility to developers, and I currently don't intend to change this part. It's really cool that packages like `next-intl-split` can exist for developers who really like this approach! 🙂

**Reply by SalahAdDin** [reply]
> `next-intl` is currently agnostic to how you load your messages: It can be a single file, you can fetch from an endpoint, use multiple files etc. I think this approach gives a lot of flexibility to developers, and I currently don't intend to change this part. It's really cool that packages like `next-intl-split` can exist for developers who really like this approach! 🙂

You have an excellent point.

**Reply by HPouyanmehr** [reply]
> @HPouyanmehr How can it be used with `next-config.mjs`

From v1.2.2 of `next-intl-split` you can also use that with `next.config.mjs`

**Reply by SalahAdDin** [reply]
> > @HPouyanmehr How can it be used with `next-config.mjs`
> 
> From v1.2.2 of `next-intl-split` you can also use that with `next.config.mjs`

what about typescript?

**Reply by HPouyanmehr** [reply]
> > > @HPouyanmehr How can it be used with `next-config.mjs`
> > 
> > 
> > From v1.2.2 of `next-intl-split` you can also use that with `next.config.mjs`
> 
> what about typescript?

You can now have the `TypeScript` support as well, from `v1.2.4`. Check [this section](https://github.com/HPouyanmehr/next-intl-split) for more information.

**Reply by SalahAdDin** [reply]
> > > > @HPouyanmehr How can it be used with `next-config.mjs`
> > > 
> > > 
> > > From v1.2.2 of `next-intl-split` you can also use that with `next.config.mjs`
> > 
> > 
> > what about typescript?
> 
> You can now have the `TypeScript` support as well, from `v1.2.4`. Check [this section](https://github.com/HPouyanmehr/next-intl-split) for more information.

So, the package is like a plugin. Right?

**Reply by HPouyanmehr** [reply]
> > > > > @HPouyanmehr How can it be used with `next-config.mjs`
> > > > 
> > > > 
> > > > From v1.2.2 of `next-intl-split` you can also use that with `next.config.mjs`
> > > 
> > > 
> > > what about typescript?
> > 
> > 
> > You can now have the `TypeScript` support as well, from `v1.2.4`. Check [this section](https://github.com/HPouyanmehr/next-intl-split) for more information.
> 
> So, the package is like a plugin. Right?

Yes, of course, using the package, you'll be able to separate the translation files in your desired way with less effort, but at the end of the day, the translations will merge, and for production, you'll have just one JSON file per locale. 

So, this is not a replacement; it's just a tiny plugin to improve the DX, especially if you're a fan of modularity.

**Comment by denisb97** [comment]
For those who are looking for an automated way to add multiple dictionary files without the need of adding the import path manually here's the solution which works for us perfectly:

- update the `i18n/request.ts` with the snippet below
- script assumes that you have the `global-dictionary` folder (you can adjust/remove that if needed)
- the script scans for `dictionary` folders throughout your project 

The comments are self-explanatory though :) 

```
/**
 * Translation System for Next.js with next-intl
 * ==============================================
 *
 * This file manages the automatic discovery and loading of translation files across the project.
 * Instead of manually specifying each translation file, this system scans the project
 * for all dictionary folders and automatically imports and merges translation files.
 *
 * How It Works:
 * -------------
 * 1. The system gets the user's locale from the getUserLocale service
 * 2. It loads the main dictionary file from /dictionary/{locale}.json
 * 3. It scans the entire project for folders named "dictionary"
 * 4. For each dictionary folder found, it checks if a file exists for the current locale
 * 5. If the file exists, it dynamically imports and merges the translations
 *
 * Creating Dictionary Folders:
 * ---------------------------
 * To add translations for a specific feature or route:
 *
 * 1. Create a folder named "dictionary" in the relevant feature directory:
 *    Example: app/(protected-pages)/(master-data)/your-feature/dictionary/
 *
 * 2. Add translation files named after the locale:
 *    - en.json for English
 *    - de.json for German
 *    - etc.
 *
 * 3. The files will be automatically discovered and loaded - no need to
 *    manually import them in this file!
 *
 * Example Dictionary Structure:
 * ----------------------------
 * /global-dictionary/en.json                           <- Main translations
 * /app/(protected-pages)/users/dictionary/en.json  <- User-specific translations
 * /app/(protected-pages)/orders/dictionary/en.json <- Order-specific translations
 */

import { getUserLocale } from "@/services";
import { getRequestConfig } from "next-intl/server";
import fs from "fs";
import path from "path";

/**
 * Recursively find all dictionary files for a specific locale
 *
 * @param dir - The directory to search in
 * @param locale - The locale to look for (e.g., 'en', 'de')
 * @param result - Array to collect paths to dictionary folders
 * @returns Array of paths to dictionary folders containing the locale file
 */
const findDictionaryFiles = (
  dir: string,
  locale: string,
  result: string[] = []
): string[] => {
  const files = fs.readdirSync(dir);

  for (const file of files) {
    const fullPath = path.join(dir, file);
    const stat = fs.statSync(fullPath);

    if (stat.isDirectory()) {
      // If there's a dictionary folder, check if it contains our locale file
      if (file === "dictionary") {
        const localePath = path.join(fullPath, `${locale}.json`);
        if (fs.existsSync(localePath)) {
          result.push(fullPath);
        }
      } else {
        // Continue searching in subdirectories
        findDictionaryFiles(fullPath, locale, result);
      }
    }
  }

  return result;
};

export default getRequestConfig(async ({ locale }) => {
  // Ensure locale is always a string by providing a fallback
  const userLocale = (await getUserLocale()) || locale || "en";

  // Start with the main dictionary from the root dictionary folder
  let messages = (await import(`../global-dictionary/${userLocale}.json`))
    .default;

  // Find all dictionary folders in the project
  const projectRoot = path.resolve(".");
  const dictionaryPaths = findDictionaryFiles(projectRoot, userLocale);

  // Import and merge all dictionary files
  for (const dictPath of dictionaryPaths) {
    // Skip the main dictionary we already imported
    if (dictPath !== path.join(projectRoot, "dictionary")) {
      const relativePath = path.relative(projectRoot, dictPath);
      const modulePath = `../${relativePath}/${userLocale}.json`;
      try {
        const moduleMessages = (await import(modulePath)).default;
        messages = { ...messages, ...moduleMessages };
      } catch (error) {
        console.error(`Failed to import dictionary from ${modulePath}`, error);
      }
    }
  }

  return {
    locale: userLocale,
    messages,
  };
});

```
